File: generator/src/main/groovy/ch/bailu/gtk/tag/NamespaceTag.java
package ch.bailu.gtk.tag;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;


public class NamespaceTag extends NamedTag {
    public NamespaceTag(Tag parent) {
        super(parent);
    }


    private List<NamedTag> includes = new ArrayList<>();
    private List<MethodTag> functions = new ArrayList<>();

    private NamedTag _package = new NamedTag(this);

    @Override
    public void started() throws IOException {
        getBuilder().buildNamespaceStart(this);
    }


    @Override
    public void end() throws IOException {
        getBuilder().buildNamespaceEnd(this);
    }

    @Override
    public Tag getChild(String name, String prefix) {
        if ("class".equals(name)) {
            return new StructureTag(this, name);
        }

        if ("record".equals(name)) {
            return new StructureTag(this, name);
        }

        if ("interface".equals(name)) {
            return new StructureTag(this, name);
        }

        if ("enumeration".equals(name)) {
            return new EnumerationTag(this);
        }

        if ("bitfield".equals(name)) {
            return new EnumerationTag(this);
        }

        if ("alias".equals(name)) {
            return new AliasTag(this);
        }

        if ("function".equals(name)) {
            MethodTag f = new MethodTag(this);
            functions.add(f);
            return f;
        }

        return ignore();
    }

    public List<NamedTag> getIncludes() {
        return includes;
    }

    public List<MethodTag> getFunctions() {
        return functions;
    }

    public String getPackage() {
        return _package.getName();
    }

    public Tag addInclude(NamedTag include) {
        includes.add(include);
        return include;
    }

    public Tag addPackage(NamedTag pkg) {
        _package = pkg;
        return pkg;
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/parser/tag/NamespaceTag.kt
package ch.bailu.gtk.parser.tag

import java.io.IOException
import java.util.*

class NamespaceTag(parent: TagWithParent): NamedWithDocTag(parent) {

    private val includes: MutableList<NamedWithDocTag> = ArrayList()
    private val functions: MutableList<MethodTag> = ArrayList()
    private val constants: MutableList<ParameterTag> = ArrayList()

    private var packageTag = NamedWithDocTag(this)

    @Throws(IOException::class)
    override fun started() {
        getBuilder().buildNamespaceStart(this, getNamespaceConfig())
    }


    @Throws(IOException::class)
    override fun end() {
        getBuilder().buildNamespaceEnd()
    }

    override fun getChild(name: String): TagWithParent {
        if ("class" == name) {
            return StructureTag(this, name)
        }
        if ("record" == name) {
            return StructureTag(this, name)
        }
        if ("union" == name) {
            return StructureTag(this, name)
        }
        if ("interface" == name) {
            return StructureTag(this, name)
        }
        if ("enumeration" == name) {
            return EnumerationTag(this)
        }
        if ("constant" == name) {
            val c = ParameterTag(this)
            constants.add(c)
            return c
        }
        if ("bitfield" == name) {
            return EnumerationTag(this)
        }
        if ("alias" == name) {
            return AliasTag(this)
        }
        if ("function" == name) {
            val f = MethodTag(this)
            functions.add(f)
            return f
        }
        if ("callback" == name) {
            return CallbackTag(this)
        }
        return super.getChild(name)
    }

    fun getIncludes(): List<NamedWithDocTag> {
        return includes
    }

    fun getFunctions(): List<MethodTag> {
        return functions
    }

    fun getConstants(): List<ParameterTag> {
        return constants
    }

    fun getPackage(): String {
        return packageTag.getName()
    }

    fun addInclude(include: NamedWithDocTag): TagWithParent {
        includes.add(include)
        return include
    }

    fun addPackage(pkg: NamedWithDocTag): TagWithParent {
        packageTag = pkg
        return pkg
    }
}
